bika/lims/browser/widgets/analysisspecificationwidget.py

```diff
@@ -35,152 +50,216 @@ def __init__(self, context, request, fieldvalue=[], allow_edit=True):
-            if service.getAttachmentOption() == 'r':
-                after_icons += "<img\
-                src='%s/++resource++bika.lims.images/attach_reqd.png'\
-                title='%s'>"%(self.context.absolute_url(),
-                              _("Attachment required"))
-            if service.getAttachmentOption() == 'n':
-                after_icons += "<img\
-                src='%s/++resource++bika.lims.images/attach_no.png'\
-                title='%s'>"%(self.context.absolute_url(),
-                              _('Attachment not permitted'))
-
-            # TRICK for AS keyword retrieval on form POST
-            after_icons += '<input type="hidden" name="keyword.%s:records"\
-            value="%s"></input>' % (service.UID(), service.getKeyword())
-
-            state = workflow.getInfoFor(service, 'inactive_state', '')
-            unit = service.getUnit()
-            unitspan = unit and \
-                "<span class='discreet'>%s</span>" % service.getUnit() or ''
-            percspan = "<span class='discreet'>%</span>"
-
-            item = {
-                'obj': service,
-                'id': service.getId(),
-                'uid': service.UID(),
-                'keyword': service.getKeyword(),
-                'title': service.Title(),
-                'category': cat,
-                'selected': service.getKeyword() in self.specsresults.keys(),
-                'type_class': 'contenttype-ReferenceResult',
-                'url': service.absolute_url(),
-                'relative_url': service.absolute_url(),
-                'view_url': service.absolute_url(),
-                'service': service.Title(),
-                'min': specresults.get('min', ''),
-                'max': specresults.get('max', ''),
-                'warn_min': specresults.get('warn_min', ''),
-                'warn_max': specresults.get('warn_max', ''),
-                'hidemin': specresults.get('hidemin',''),
-                'hidemax': specresults.get('hidemax',''),
-                'rangecomment': specresults.get('rangecomment', ''),
-                'replace': {},
-                'before': {},
-                'after': {'service':after_icons,
-                          'min':unitspan,
-                          'max':unitspan,
-                          'warn_min':unitspan,
-                          'warn_max':unitspan},
-                'choices':{},
-                'class': "state-%s" % (state),
-                'state_class': "state-%s" % (state),
-                'allow_edit': ['min', 'max', 'warn_min', 'warn_max', 'hidemin',
-                               'hidemax', 'rangecomment'],
-            }
-            items.append(item)
 
-        self.categories.sort()
-        for i in range(len(items)):
-            items[i]['table_row_class'] = "even"
+        services = self.get_services()
+        for num, service in enumerate(services):
+            item = self.folderitem(service, {}, num)
+            if item:
```

Comment:
As far as I see, `folderitem` will always return a dict, never empty or None, so you could omit this conditional